processor/src/chiplets/ace/mod.rs

```diff
@@ -0,0 +1,292 @@
+
+        let mut gen_trace: [Vec<Felt>; ACE_CHIPLET_NUM_COLS] = (0..ACE_CHIPLET_NUM_COLS)
+            .map(|_| vec![ZERO; self.trace_len()])
+            .collect::<Vec<_>>()
+            .try_into()
+            .expect("failed to convert vector to array");
+
+        let mut offset = 0;
+        for eval_ctx in self.circuit_evaluations.into_values() {
+            eval_ctx.fill(offset, &mut gen_trace);
+            offset += eval_ctx.num_rows();
+            let section = AceSection::from_evaluation_context(&eval_ctx);
+            self.sections_info.push(section);
+        }
+
+        for (out_column, column) in trace.columns().zip(gen_trace) {
+            out_column.copy_from_slice(&column);
+        }
+
+        self.sections_info
+    }
+
+    pub(crate) fn add_eval_context(&mut self, clk: RowIndex, eval_context: EvaluationContext) {
+        self.circuit_evaluations.insert(clk.as_u32(), eval_context);
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct AceSection {
+    ctx: u32,
+    clk: u32,
+    num_vars: u32,
+    num_evals: u32,
+}
+
+impl AceSection {
+    pub fn clk(&self) -> u32 {
+        self.clk
+    }
+
+    pub fn ctx(&self) -> u32 {
+        self.ctx
+    }
+
+    pub fn num_vars(&self) -> u32 {
+        self.num_vars
+    }
+
+    pub fn num_evals(&self) -> u32 {
+        self.num_evals
+    }
+
+    fn from_evaluation_context(eval_ctx: &EvaluationContext) -> AceSection {
+        AceSection {
+            ctx: eval_ctx.ctx(),
+            clk: eval_ctx.clk(),
+            num_vars: eval_ctx.num_read_rows(),
+            num_evals: eval_ctx.num_eval_rows(),
+        }
+    }
+}
+#[derive(Debug, Default)]
+pub struct AceHints {
+    offset_chiplet_trace: usize,
+    pub sections: Vec<AceSection>,
+}
+
+impl AceHints {
+    pub fn new(offset_chiplet_trace: usize, sections: Vec<AceSection>) -> Self {
+        Self { offset_chiplet_trace, sections }
+    }
+
+    pub(crate) fn offset(&self) -> usize {
+        self.offset_chiplet_trace
+    }
+
+    pub(crate) fn build_divisors<E: FieldElement<BaseField = Felt>>(
+        &self,
+        main_trace: &MainTrace,
+        alphas: &[E],
+    ) -> Vec<E> {
+        let num_fractions = self.num_fractions();
+        let mut total_values = vec![E::ZERO; num_fractions];
+        let mut total_inv_values = vec![E::ZERO; num_fractions];
+
+        let mut chiplet_offset = self.offset_chiplet_trace;
+        let mut values_offset = 0;
+        let mut acc = E::ONE;
+        for section in self.sections.iter() {
+            let clk = section.clk();
+            let ctx = section.ctx();
+
+            let values = &mut total_values[values_offset
+                ..values_offset + NUM_ACE_LOGUP_FRACTIONS_READ * section.num_vars() as usize];
+            let inv_values = &mut total_inv_values[values_offset
+                ..values_offset + NUM_ACE_LOGUP_FRACTIONS_READ * section.num_vars() as usize];
```

Comment:
Small nit, but maybe not much clearer (up to you)
```suggestion
            let values_range =
                range(values_offset, NUM_ACE_LOGUP_FRACTIONS_READ * section.num_vars() as usize);
            let values = &mut total_values[values_range];
            let inv_values_range =
                range(values_offset, NUM_ACE_LOGUP_FRACTIONS_READ * section.num_vars() as usize);
            let inv_values = &mut total_inv_values[inv_values_range];
```